### ISSUE TITLE
[Bug]: Delay on informational views to get updated

### ISSUE THREAD

**Issue by henriquegelio** [timescale/timescaledb#5315]
### What type of bug is this?

Other

### What subsystems and features are affected?

Other

### What happened?

Hi,

I'm currently experiencing a delay on the informational views after decompressing a table.

I have a table with some rows, which is compressed, and then I decompress it manually doing:

`select decompress_chunk(i, true) from show_chunks('table') as i`

and:

```
ALTER TABLE table
SET (timescaledb.compress = false)
```

After disabling compression on the table, I try to consume the chunks and compression_settings views, and they say that the hypertable 'table' is still compressed, i.e, the compression status is still 'Compressed'. It only gets 'Uncompressed' after a few seconds (~7, 8 seconds).

All this is done via JDBC using java. I'm not sure is this could be a JDBC issue, because when I do the same process using pgAdmin, I couldn't notice this issue.

We have already tried to use exclusive connections on JDBC, but with no success.

Do you have any idea on what might be happening and how to solve this issue?

Thank you in advance and sorry for the label Bug. Not sure is this is a bug but couldn't find any better option.

### TimescaleDB version affected

2.9.0

### PostgreSQL version used

14

### What operating system did you use?

Ubuntu 20

### What installation method did you use?

Docker

### What platform did you run on?

On prem/Self-hosted

### How can we reproduce the bug?

```bash
- Create an hypertable
- Compress it
- Decompress it
- Consume the views
```

**Comment by kgyrtkirk** [comment]
> All this is done via JDBC using java. I'm not sure is this could be a JDBC issue, because when I do the same process using pgAdmin, I couldn't notice this issue.

I've tried to repro it via psql:
```
psql -a -v ON_ERROR_STOP=1 tsdb <<EOF

alter table readings set (timescaledb.compress=true);
select compress_chunk(i, true) from show_chunks('readings') as i;
select * from timescaledb_information.compression_settings ;
select decompress_chunk(i, true) from show_chunks('readings') as i;
alter table readings set (timescaledb.compress=false);
select * from timescaledb_information.compression_settings;

EOF
```
I didn't seen any issues.

I wonder:

* you would normally change the compression and look at the catalog tables from the java application?
   * are you using the same jdbc  connection?
* Could you tell a bit more about what java libraries are you using? I'm wondering if you have some connection pooling lib - those libraries tend to also have a caching layer.
  Note: my experience is that they usually also provide some mechanism to get direct access to an underlying connection directly - from the above I guess you've already tried that?

**Comment by henriquegelio** [comment]
Hi @kgyrtkirk ,

Thanks to your answer, I've figured out what's actually happening.

In fact I'm using a pooling mechanism, but I've ensured I'm getting an exclusive connection to the database.
The issue is not actually the views, they're showing the correct result.

The delay is being caused by the background jobs API, which I'm using to compress/decompress hypertables.

To give you more details, some days ago I've opened an [issue](https://github.com/timescale/timescaledb/issues/5188) wondering how we could achieve compression and decompression in an asynchronous manner.

By asynchronous, I mean registering a job to run in background and return immediately, without waiting for the compression/decompression to end. 

The suggested approach, which we were already using, is to have two separate UDA's. One for compression, and one for decompression. Those UDA's do exactly the same process described above, that is:

compression UDA:
`SELECT compress_chunk(i, true) FROM show_chunks(config->>'hypertable') AS i`

decompression UDA:
`SELECT decompress_chunk(i, true) FROM show_chunks(config->>'hypertable') AS i`

Every time a compression/decompression request arrives, we register a job with the `add_job()` API and delete it after it ends. Actually, instead of deleting them, we are setting the scheduled attribute to false, as suggested by @konskov.

Well.. this is working.. but the delay is happening when we have two jobs for the same hypertable being registered one after another, and I've detected this behavior in a test.

What the test does:
- Creates a database
- Creates an hypertable
- Insert two rows in this hypertable
- Enable compression
- Register a compression job (via add_job)
- Waits for 1500ms (more than enough for the hypertable to be compressed)
- Consumes timescaledb_information.chunks to ensure all chunks are compressed. (they are)
- Consumes timescaledb_information.job_stats to ensure the registered job has a 'Succeeded' status (it has).
- Register a new decompression job (via add_job)
- Waits for the same 1500ms
- Consumes timescaledb_information.chunks to ensure all chunks are decompressed (and here, the test fails).
- Consumes timescaledb_information.job_stats to ensure the job has a 'Succeeded' status (fails, it's actually 'Scheduled').

This test only passes if I wait for about 6 or 7 seconds after registering the second job.
My initial suspicion is what motivated the opening of this issue, because I thought the views were not showing the expected compression result.

With further investigation, I've noticed that the second job registered with the add_job() API **does not get run immediately**. It takes about 5 seconds after the first job ends until it starts running, even though when registering a job I'm setting the initial_start attribute to now().

Is this an expected behavior?

Important to say that these are the only two jobs being registered so far (apart from telemetry report and retention policy, which are automatically created), so there isn't any job running after the first one finishes.

Ah, I'm also setting "timescaledb.bgw_launcher_poll_time=10" before initializing the timescaledb_container, due to another [issue](https://github.com/timescale/timescaledb/issues/4217) I've also reported some time ago because the jobs were only starting after one minute after the creation of the database, and @konskov happily provided us with this feature (GUC variable).

Any ideas on what might be going on?

Thanks once again.

**Comment by kgyrtkirk** [comment]
Awesome - that sounds like a reasonable explanation :)

I think `timescaledb.bgw_launcher_poll_time=10` should already make bg job starts pretty much instant with ~10ms delay.

I think the best would be to check&wait until the job finishes (say for at most 30 seconds); (I guess you might need to also  check last_successful_finish) and proceed with other checks only after then.
As a matter of fact if you could check after how much delay the job transitions into `Running` ; and back to `Scheduled` might help decide if this is a runner issue

**Comment by henriquegelio** [comment]
> Awesome - that sounds like a reasonable explanation :)
> 
> I think `timescaledb.bgw_launcher_poll_time=10` should already make bg job starts pretty much instant with ~10ms delay.
> 
> I think the best would be to check&wait until the job finishes (say for at most 30 seconds); (I guess you might need to also check last_successful_finish) and proceed with other checks only after then. As a matter of fact if you could check after how much delay the job transitions into `Running` ; and back to `Scheduled` might help decide if this is a runner issue

Hi @kgyrtkirk .

To give you a feedback, we did exactly that (check and wait until the job finishes).

But as far as i could tell, this looks like a runner issue indeed.
When registering more than one job via add_job API, the second job takes between 4000ms and 5000ms before transitioning into "Running". Then, ~1000ms to get back to "Scheduled".
The second delay is expected, as it's the time the job takes to run. The first one, doesn't.

As a matter of fact, I've tried different types of jobs (different UDA's and even policies). In all cases, the second job delays between 4000 ~ 5000ms to transition into Running, even though there is no other job running (theoretically, the job pool is empty).

This makes me guess that it does look like a runner issue. What do you think?

Ah, I'm using fixed_schedule as FALSE in ass jobs, due to another issue I've opened some time ago. It was already fixed in 2.10 (thanks for that) but I haven't been able to update yet.

**Comment by jnidzwetzki** [comment]
Hello @henriquegelio,

Thanks for providing all the information. I would like to re-create the issue in my local setup for further analysis. Could you share the SQL commands that you have used to add the jobs and to query the view? This would help me a lot in reproducing this issue.
